test/subscription_map.cpp

```diff
@@ -0,0 +1,284 @@
+    for(auto const &i: values) {
+        handles.push_back(map.insert(i, i));
+    }
+
+    for(auto const &i: handles) {
+        BOOST_TEST(map.size() != 0);
+        BOOST_TEST(map.erase(i) == 1);
+    }
+
+    BOOST_TEST(map.size() == 1);
+
+}
+
+BOOST_AUTO_TEST_CASE( TestMultipleSubscription ) {
+    std::string text = "example/test/A";
+
+    multiple_subscription_map<std::string> map;
+
+    map.insert("a/b/c", "123");
+    map.insert("a/b", "123");
+
+    map.erase("a/b/c", "123");
+    BOOST_TEST(map.size() != 1);
+
+    map.erase("a/b", "123");
+    BOOST_TEST(map.size() == 1);
+
+    std::vector<std::string> values = {
+        "example/test/A", "example/+/A", "example/#", "#"
+    };
+
+    // Add some duplicates and overlapping paths
+    map.insert(values[0], values[0]);
+    BOOST_TEST(map.insert(values[0], values[0]).second == false);
+    BOOST_TEST(map.insert(values[0], "blaat").second == true);
+
+    map.erase(values[0], "blaat");
+    BOOST_TEST(map.size() != 1);
+
+    map.erase(values[0], values[0]);
+    BOOST_TEST(map.size() == 1);
+
+    // Perform test again but this time using handles
+    map.insert(values[0], values[0]);
+    BOOST_TEST(map.insert(map.lookup(values[0]), values[0]).second == false);
+    BOOST_TEST(map.insert(map.lookup(values[0]), "blaat").second == true);
+
+    map.erase(map.lookup(values[0]), "blaat");
+    BOOST_TEST(map.size() != 1);
+
+    map.erase(map.lookup(values[0]), values[0]);
+    BOOST_TEST(map.size() == 1);
+
+    for(auto const &i: values) {
+        map.insert(i, i);
+    }
+
+    // Attempt to remove entry which has no value
+    BOOST_TEST(map.erase("example", "example") == 0);
+    BOOST_TEST(map.erase(map.lookup("example"), "example") == 0);
+    BOOST_TEST(map.erase("example", "example") == 0);
+    BOOST_TEST(map.erase(map.lookup("example"), "example") == 0);
+
+    BOOST_TEST(map.lookup(values[0]).size() == 3);
+    BOOST_TEST(map.handle_to_subscription(map.lookup(values[0])) == values[0]);
+
+    std::vector<std::string> matches;
+    map.find("example/test/A", [&matches](std::string const &a) {
+        matches.push_back(a);
+    });
+    BOOST_TEST(matches.size() == 4);
+
+    matches = {};
+    map.find("hash_match_only", [&matches](std::string const &a) {
+        matches.push_back(a);
+    });
+    BOOST_TEST(matches.size() == 1);
+
+    matches = {};
+    map.find("example/hash_only", [&matches](std::string const &a) {
+        matches.push_back(a);
+    });
+    BOOST_TEST(matches.size() == 2);
+
+    matches = {};
+    map.find("example/plus/A", [&matches](std::string const &a) {
+        matches.push_back(a);
+    });
+    BOOST_TEST(matches.size() == 3);
+
+    BOOST_TEST(map.erase("non-existent", "non-existent") == 0);
+
+    for(auto const &i: values) {
+        BOOST_TEST(map.size() != 0);
+        BOOST_TEST(map.erase(i, i) == 1);
+    }
+
+    BOOST_TEST(map.size() == 1);
+}
+
+
+
+BOOST_AUTO_TEST_CASE(retained_topic_map_test) {
```

Comment:
Could you separate the test for retaind_topic_map to new file `retained_topic_map.cpp` ?